pkg/raft/rawnode.go

```diff
@@ -605,7 +634,22 @@ func (rn *RawNode) ReportUnreachable(id pb.PeerID) {
 	_ = rn.raft.Step(pb.Message{Type: pb.MsgUnreachable, From: id})
 }
 
-// ReportSnapshot reports the status of the sent snapshot.
+// ReportSnapshot reports the status of the snapshot sent to the given peer.
+//
+// Any failure in sending a snapshot (e.g. while streaming it from leader to
+// follower) must be reported to the leader with SnapshotFailure.
+//
+// When the leader sends a snapshot to a peer, it pauses log replication until
+// this peer can apply the snapshot and advance its state. If the peer can't do
+// that, e.g. due to a crash, it could end up in a limbo, never getting any
+// updates from the leader. Therefore, it is crucial that the app catches any
+// failure in sending a snapshot and reports it back to the leader, so that the
+// log replication probing can resume. In case of uncertainty, the app must err
+// on the side of reporting SnapshotFailure.
+//
+// Successful snapshot sends should be reported with SnapshotFinish. It is less
+// crucial to do this, because a successful snapshot is also reported via a
+// MsgAppResp message stepped into this leader.
```

Comment:
Less crucial, but still crucial, right? The follower emits a one-off MsgAppResp but if that gets dropped, we really do rely on the `ReportSnapshot` call at the leader. (In our case the raft snapshot queue would also retry eventually, but the recipient may then discard the snapshot if the index hasn't changed, and not even emit another MsgAppResp).

Reply:
Reworded a bit.